python_modules/libraries/dagster-k8s/dagster_k8s/launcher.py

```diff
@@ -393,10 +400,44 @@ def check_run_worker_health(self, run: PipelineRun):
             job = self._batch_api.read_namespaced_job(namespace=self.job_namespace, name=job_name)
         except Exception:
             return CheckRunHealthResult(
-                WorkerStatus.UNKNOWN, str(serializable_error_info_from_exc_info(sys.exc_info()))
+                WorkerStatus.UNKNOWN,
+                f"Error checking Kubernetes Job: {str(serializable_error_info_from_exc_info(sys.exc_info()))}",
             )
         if job.status.failed:
-            return CheckRunHealthResult(WorkerStatus.FAILED, "K8s job failed")
+            return CheckRunHealthResult(
+                WorkerStatus.FAILED, "Kubernetes Job failed due to a Pod failure"
+            )
         if job.status.succeeded:
-            return CheckRunHealthResult(WorkerStatus.SUCCESS)
-        return CheckRunHealthResult(WorkerStatus.RUNNING)
+            return CheckRunHealthResult(WorkerStatus.SUCCESS, "Kubernetes Job succeeded")
+
+        pods = self._core_api.list_namespaced_pod(
+            namespace=self.job_namespace, label_selector=f"job-name={job_name}"
+        ).items
+
+        if not pods:
+            return CheckRunHealthResult(
+                WorkerStatus.WAITING, "Kuberetes Job does not have a Pod yet"
+            )
+
+        # NOTE: There's a chance of 2 pods if a node was shutdown. We'll need to handle that.
+        pod = pods[0]
+
+        if pod.status.container_statuses:
+            container_statuses = pod.status.container_statuses[0]
+            if container_statuses.state.waiting:
+                if container_statuses.state.waiting.reason in [
+                    KubernetesWaitingReasons.ErrImagePull,
+                    KubernetesWaitingReasons.ImagePullBackOff,
+                    KubernetesWaitingReasons.ErrImageNeverPull,
+                ]:
+                    return CheckRunHealthResult(
+                        WorkerStatus.FAILED,
+                        f"Kubernetes Job failed to pull image: {container_statuses.state.waiting.reason}",
+                    )
```

Comment:
I thought we determined that there's a real possibility that it can recover from this? that's what prompted us to add an 'imagePullGracePeriod' elsewhere

Reply:
That's a good point. We ran into that when loading the workspace just after pushing the image right? I think in all normal deployments the image will be fully available by the time the run gets launched, since it was needed for the workspace already

Reply:
We are dipping in to a class of things which can recover though (along with future cases like missing a secret, etc.). I could imagine us offering a toggle for failing the run on these